nemo_aligner/algorithms/spin.py

```diff
@@ -195,19 +196,23 @@ def train_single_step(self, global_batch):
         return loss_mean, {**metrics, **trainer_metrics}
 
     @torch.no_grad()
-    def get_generations(self, batch):
+    def get_generations(self, list_of_batches):
         self.model.prepare_for_inference()
 
-        prompt_lengths = batch["prompt_lengths"]
+        prompt_lengths = torch.cat([b["prompt_lengths"] for b in list_of_batches], dim=0)
         batch_max_length = prompt_lengths.max().item()
         max_possible_length = min(self.model.cfg.encoder_seq_length, batch_max_length + self.max_gen_seq_len)
         # in case the prompt length exceeds encoder_seq_length - max_gen_seq_len, we need to truncate how many
         # tokens we are allowed to generate such that we never exceed encoder_seq_length, otherwise you will get
         # errors inside model.generate()
         adj_generation_length = min(self.max_gen_seq_len, self.model.cfg.encoder_seq_length - batch_max_length)
 
-        prompt_tokens = batch_pad_to_fixed_len(
-            batch["prompts_only"], max_possible_length, pad_token=self.model.tokenizer.eos_id
+        prompt_tokens = torch.cat(
+            [
+                batch_pad_to_fixed_len(b["prompts_only"], max_possible_length, pad_token=self.model.tokenizer.eos_id)
+                for b in list_of_batches
+            ],
+            dim=0,
```

Comment:
We probably can, but is it really so bad to just do it like I have it?

Reply:
Actually, it does not appear possible to do this, because `list_of_batches` is jagged, so you cannot torch.stack its contents for passing to batch_pad_to_fixed_len.

Similarly, you cannot do: batch_pad_to_fixed_len(itertools.chain(list_of_batches), max_possible_length, pad_token=self.model.tokenizer.eos_id) as you get an error `Tensors must have same number of dimensions: got 2 and 1`

Reply:
Hmm not sure I follow but not a big deal, it was just a minor point to save a few lines of code for readability => resolving.